### ISSUE TITLE
Layer-2 Flat networking support in Rancher cattle

### ISSUE THREAD

**Issue by niusmallnan** [rancher/rancher#8673]
I have updated network-manager, it can be able to support other CNI drivers. So Flat network support can be achieved. #8535 

**How does Flat work?**
![image](https://cloud.githubusercontent.com/assets/998984/25739860/58a357c2-31b6-11e7-8bec-1a1b3198f2b8.png)
Notice:
1. Container network and eth0 in the same network
2. Move eth0 IP setting to flat-br, eth0 needs to be bound on the bridge flat-br. This part will be executed automatically

**For network-manager version requirements:**
As previously mentioned, make sure that the version you are using includes this commit https://github.com/rancher/plugin-manager/commit/6512bf77f38f3ed6751644aca8165e311634881b
Test image: **niusmallnan/network-manager:dev**

**For catalog item:**
There is a catalog item that is being tested.
Repo: https://github.com/niusmallnan/flatnet-catalog.git
Branch: master

**For rancher-cni-flat:**
Basically, it is based on rancher-cni-bridge. I just add a flag named isFlat and some logic.
I'm not sure if this needs to be merged into rancher-cni-bridge.
Repo: https://github.com/niusmallnan/rancher-cni-bridge
Branch: flat

Ensure that the Rancher server version is v1.6.0+

**Comment by leodotcloud** [comment]
> Move eth0 IP setting to flat-br, eth0 needs to be bound on the bridge flat-br. This part will be executed automatically

@niusmallnan Did you mean this has to be done outside of Rancher manually by a System admin?

**Comment by niusmallnan** [comment]
Use the following cni configuration:
```
'10-rancher-flat.conf':
          name: rancher-cni-network
          type: rancher-bridge
          bridge: ${FLAT_BRIDGE}
          bridgeSubnet: ${SUBNET}
          bridgeIP: '__host_interface__: ${FLAT_BRIDGE}'
          logToFile: /var/log/rancher-cni.log
          isDebugLevel: ${RANCHER_DEBUG}
          hairpinMode: {{ .Values.RANCHER_HAIRPIN_MODE }}
          promiscMode: {{ .Values.RANCHER_PROMISCUOUS_MODE }}
          hostNat: {{ .Values.HOST_NAT }}
          mtu: ${MTU}
          ipam:
            type: rancher-cni-ipam
            logToFile: /var/log/rancher-cni.log
            isDebugLevel: ${RANCHER_DEBUG}
            subnetPrefixSize: /${SUBNET_PREFIX_SIZE}
            routes:
            - dst: 0.0.0.0/0
              gw: ${GATEWAY}
            - dst: 169.254.169.250/32
              gw: '__host_interface__: ${FLAT_BRIDGE}'
```
Replace the `__host_interface__` value in the plugin-manager.

**Comment by leodotcloud** [comment]
@niusmallnan can you please update the catalog, necessary repos with the latest network-manager and cni-driver changes? If you need any changes in the CNI plugins, can you please make PRs for them to review?

**Comment by leodotcloud** [comment]
@niusmallnan One important point to consider for this feature is the task of moving the interface to the flat bridge. Who should be responsible? Initial discussions pointed out that network/system administrator would take care of the step or use startup scripts. If we were to automate this inside rancher, probably we might need a label for each host to know the interface name as this varies from one OS to another. For example on RHEL or CentOS, the network interfaces are not named `eth0`, they look something like `ens21`.

**Comment by niusmallnan** [comment]
@leodotcloud 
`moving the interface to the flat bridge`
Users are free to choose, whether automation will be an option.

The default value for `interface` comes from the option of catalog item, if the user needs to set a different interface for a host, it can be implemented via host label.

**Comment by leodotcloud** [comment]
Looking back:
```
bridgeIP: '__host_interface__: ${FLAT_BRIDGE}'
              gw: '__host_interface__: ${FLAT_BRIDGE}'

```
These two lines are incorrect, we don't need to introduce new keyword `__host_interface__`.

> Users are free to choose, whether automation will be an option.
Sounds good. This should be a boolean question in catalog item.

> The default value for interface comes from the option of catalog item, if the user needs to set a different interface for a host, it can be implemented via host label.
Sounds good. If the host label is found use that, else use the default value from the catalog.

> BTW, is there any way to disable fastpath in rancher-cni-bridge?
Currently there is none. One way I could think of is to introduce a new boolean argument in `plugin-manager` (say `--disable-fastpath`) which will remove the label IP label from the docker inspect before invoking cniglue. Here is the line: https://github.com/rancher/plugin-manager/blob/master/network/manager.go#L61. 
Pseudo code:
```
if (os.env(DISABLE_FASTPATH) {
    delete(inspect.label[io.rancher.container.ip]
}
```

**Comment by leodotcloud** [comment]
Food for thought: What if we use fastpath if there is no ipam keyword in the CNI config? Not sure if this will address all the use cases in 2.0. Documenting it here so that we can revisit this discussion again.

**Comment by yasker** [comment]
@kingsd041 Can you test this with https://github.com/rancher/rancher-catalog/pull/981 ?

**Comment by yasker** [comment]
Catalog merged.

**Comment by kingsd041** [comment]
Rancher/server: v1.6.11-rc5
Catalog: https://github.com/rancher/rancher-catalog.git, branch: v1.6-development
L2 Flat: v0.1.0
Rancher/net: v0.13.4
Docker version: 1.12.6
Provider of hosts: VirtualBox

The test is done and no new issues are found.
